### PR TITLE
make default values stored by ERR_get_error_all() and friends more co…

### PR DESCRIPTION
**Pull request by DDvO**
When chasing a regression in an error reporting test of our CMP contribution 
I came again across the issue that the documentation of `ERR_get_error_all()` and friends
does not state that they touch storage locations for which pointers are given only on success and 
does not state which values are stored in case no respective data (e.g., line number or function name) is available. Therefore it remains unclear if callers should initialize the respective locations beforehand and how they should detect the absence of a an actual value.

Moreover, the default pointer values provided by these functions in case the requested information is unset so far are rather inconsistent: `""` vs. `"NA"` vs. `"N/A"`.

This PR clarifies these things in the documentation and suggests a consistent default for pointer values: `NULL`. I chose `NULL` rather than `""` because the latter could be more easily confused with a legitimate value. If for compatibility with existing usage a proper pointer is needed (though I did not notice any such need at least in the existing test suite), I'd say `""` is better than `"NA"` or any other non-empty string because such a non-empty string could collide with a real value (e.g., function name).   

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Cleaner code too.  I'm all for it.

**Comment by t8m** [comment]
As the existing code never returned NULL pointers in the file/line/func/data output values, I'd really prefer if this was kept with the current patch. Otherwise it would be a slight API break that could produce crashes in the existing third party code that uses these API calls.
I have no problem with making it all consistent and empty strings.

**Comment by t8m** [comment]
As for the documentation improvement I'd have another suggestion (which does not block this PR, but could be added to it) - it should be mentioned that the data value is valid only until any other OpenSSL API calls that can manipulate the error stack are called, because after that call the data value can be deallocated.

File: doc/man3/ERR_get_error.pod [hunk 1]
```diff
@@ -68,21 +68,29 @@ messages.
 
 ERR_get_error_line(), ERR_peek_error_line() and
 ERR_peek_last_error_line() are the same as ERR_get_error(),
-ERR_peek_error() and ERR_peek_last_error(), but they
+ERR_peek_error() and ERR_peek_last_error(), but on success they
 additionally store the file name and line number where
-the error occurred in *B<file> and *B<line>, unless these are B<NULL>.
+the error occurred in *B<file> and *B<line>, as far as they are not B<NULL>.
+An unset file name is indicated as B<NULL>.
```
Comment (t8m) [inline doc/man3/ERR_get_error.pod]: Please also change the unset values from NULL to empty strings.

**Comment by DDvO** [comment]
Thanks @t8m also for the following comment:

> As the existing code never returned NULL pointers in the file/line/func/data output values, I'd really prefer if this was kept with the current patch. Otherwise it would be a slight API break that could produce crashes in the existing third party code that uses these API calls.
> I have no problem with making it all consistent and empty strings.

I was considering this, too, as indirectly indicated in my original comments to this PR.
Still my preference is to return `NULL` (rather than `""`) in case of unset values because

* this is the most clean approach,
* so far the behavior in such cases was undefined, so any prudent application developer
will not have relied on always obtaining non-`NULL` pointers here, and
* even when passing a NULL pointer to `printf ( ...%s...)` or any similar output function, this does not result in a crash but in, e.g., `<NULL>` being output.

Are there further voices on this design decision?

**Comment by t8m** [comment]
Passing NULL pointer to %s printf format is undefined behavior. Yes, it currently does not crash with glibc but it could. 
Yes, any prudent app developer would probably take care of properly handling NULL because there was nothing said about the possibility of getting NULL in the docs. However I'd expect there definitely are applications that do not check for NULLs and blindly pass the pointers to the %s printf formats or similar APIs. There could be also app developers that looked up the source and saw that the code had explicit handling by overriding NULLs with NA or empty string.
Not having test case in the OpenSSL code that would test for this is not really indicative that much.

**Comment by DDvO** [comment]
BTW, I've just updated also the reference to ERR_GET_LIB(3) in the docs because location info in the error code meanwhile has become less usable due to the deprecation of the function code.

**Comment by DDvO** [comment]
@t8m, I tend to agree that to be on the safe side it may be better to return `""` rather than `NULL` for unset pointer values.

@levitte, since you already approved the `NULL` solution, what do you (meanwhile) think about that?

Of course, I can easily (re-)add the requested defensive checks on `NULL` in `get_error_values()`.

 Any other strong opinions in either direction?

**Comment by levitte** [comment]
> * even when passing a NULL pointer to `printf ( ...%s...)` or any similar output function, this does not result in a crash but in, e.g., `<NULL>` being output.

You can't trust this.  *Some* C libraries do this, others don't (worst case scenario, they crash).  According to the C language standard, an unexpected value or type leads to undefined behaviour.

**Comment by levitte** [comment]
I'm a bit ambivalent about the what's being discussed.  I *do* like returning NULL as an indicator of no value present, but at the same time, @t8m is entirely correct that this might be unexpected.  The more cautious approach would be to return the empty string as an indicator (i.e. change "NA" and "N/A" to "").

**Comment by DDvO** [comment]
Since nobody else gave his opinion on this matter and @levitte retracted this approval, I've changed the value returned for unset pointers from `NULL` to `""` as requested by @t8m to be on the safe side.

**Comment by DDvO** [comment]
How about merging this now?

**Comment by levitte** [comment]
Yes, thanks for the reminder

**Comment by levitte** [comment]
I squashed the commits into one, that seemed appropriate.

df0822688fc3432cf800cdc07c7f9016ea201170 Make default values by ERR_get_error_all() and friends more consistent

**Pull request closed without merging**